src/StateSubject.js

```diff
@@ -0,0 +1,61 @@
+import { Subject } from 'rxjs/Subject';
+
+// Used as a placeholder value so we can tell
+// whether or not the real state has been set,
+// even if the real state was `undefined`
+const UNSET_VALUE = {};
+
+// We can't use BehaviorSubject because when we create
+// and give state$ to your epics your reducers have not
+// yet run, so there is no initial state yet until the
+// `@@redux/INIT` action is emitted. BehaviorSubject
+// requires an initial value (which would be undefined)
+// and if an epic subscribes to it on app boot it would
+// synchronously emit that undefined value incorrectly.
+export class StateSubject extends Subject {
+  constructor(store) {
+    super();
+    // If you're reading this, keep in mind that this is
+    // NOT part of the public API and will be removed!
+    this.__store = store;
+    this._value = UNSET_VALUE;
+  }
+
+  _subscribe(subscriber) {
+    const subscription = super._subscribe(subscriber);
+    if (this._value !== UNSET_VALUE && subscription && !subscription.closed) {
+      subscriber.next(this._value);
+    }
+    return subscription;
+  }
+
+  get value() {
+    if (this._value === UNSET_VALUE) {
+      if (process.env.NODE_ENV !== 'production') {
+        require('./utils/console').warn('You accessed state$.value inside one of your Epics, before your reducers have run for the first time, so there is no state yet. You\'ll need to wait until after the first action (@@redux/INIT) is dispatched or by using state$ as an Observable.');
+      }
+      return undefined;
+    } else {
+      return this._value;
+    }
+  }
+
+  next(value) {
+    this._value = value;
+    return super.next(value);
+  }
+}
+
+StateSubject.prototype.getState = function () {
+  if (process.env.NODE_ENV !== 'production') {
+    require('./utils/console').deprecate('calling store.getState() in your Epics is deprecated and will be removed. The second argument to your Epic is now a stream of state$ (a StateSubject), instead of the store. To imperatively get the current state use state$.value instead of getState(). Alternatively, since it\'s now a stream you can compose and react to state changes.\n\n  function <T, R, S, D>(action$: ActionsObservable<T>, state$: StateSubject<S>, dependencies?: D): Observable<R>\n\nLearn more: https://goo.gl/WWNYSP');
```

Comment:
The google URL https://goo.gl/WWNYSP will need to be updated to a new entry in the docs.

Reply:
Wordsmithing on this deprecation message is welcome!

Reply:
Should `function <T, R, S, D>` be `function <T, S, D>`?

Reply:
What would be the usual pattern if people were using `getState()` often inside of epics, usually at the start? `combineLatest`?

Reply:
@rgbkrk `action$.ofType(Constant).withLatestFrom(state$)` is the pattern you'd most often use. I think we should make sure to be detailed about best Rx practices here, because combineLatest is not going to be the behavior people want, and I think it's an easy trap to fall into. You want the action$ filter source observable to throttle any further down calculations. `combineLatest` will fire when _any_ of those observables fires (if I'm understanding correctly).

Reply:
That makes total sense, you only want to react to the action coming in, not every change in the state should trigger it. I'll probably write it this way:

```js
action$.pipe(
  ofType(Constant),
  withLatestFrom(state$),
  ...
)
```

Reply:
Also I believe `<T, R, S, D>` is  the correct typings for Epics. The `R` generic parameter represents possible epic "output" actions. That feature has been asked for [in a few issues](https://github.com/redux-observable/redux-observable/issues/219). The `R` generic parameter should probably be either before or after the `D` dependency generic parameter to allow for a good use of defaults:
`<T, S, D = any, R = T>`

Which will allow most people to ignore this specific feature.

Reply:
👍  I just added a couple contrived examples to the OP. I agree the docs will need some best practices. I imagine a large majority of people will just use `state$.value` and won't have a lot of need for state as a stream. But it's handy when you do need it.

Reply:
> Should function <T, R, S, D> be function <T, S, D>?

@rgbkrk It's not decided, but @ksaldana1 is right about what I was thinking; the ability to specify different type for the output stream. That'll have to be a separate PR. For now I think I'll remove this type signature from the deprecation to unblock this PR.

> The R generic parameter should probably be either before or after the D dependency generic parameter to allow for a good use of defaults:
<T, S, D = any, R = T>

yeah I've struggled with this one. `R = T` is technically wrong as epics shouldn't output everything they input as that would be infinite recursion. Currently I'm thinking that if someone wants to type things they need to go all-in and type both input and output, with no defaults other than implicit `any` default. That way we're not encouraging incorrect things like `R = T` which could confuse people or worse hide bugs. I'll tag you both in PR for the Epic type signature stuff so we can continue the debate and refinement. I'm very happy you both are giving your thoughts.

Reply:
> `R = T` is technically wrong as epics shouldn't output everything they input as that would be infinite recursion.

This might have already been laid to rest, but I don't think that this is technically wrong. I've always seen `T` as my "universe" of action types. Anything from within that universe could come out of the `action$` stream, so it's a good input type. Re-otputting everything from the input would be infinite recursion (and bad), but this type signature doesn't imply that. It's just saying that the output could have any type from within our universe. It's a bit of a loose type, and doesn't tell us much about what the epic does, but it's not wrong.

Reply:
> It's a bit of a loose type, and doesn't tell us much about what the epic does, but it's not wrong.

Sorry, that is basically what I meant. 😄  "wrong" in a similar sense that this type signature is "wrong":

```ts
function example(): string | number {
  // always returns a number, so while it technically _could_ return a string it never will
  return 1;
}
```